video/out/gpu/context.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+#include <math.h>
+#include <assert.h>
+
+#include "config.h"
+#include "common/common.h"
+#include "common/msg.h"
+#include "options/options.h"
+#include "options/m_option.h"
+#include "video/out/vo.h"
+
+#include "context.h"
+
+extern const struct ra_ctx_fns ra_ctx_glx;
+extern const struct ra_ctx_fns ra_ctx_glx_probe;
+extern const struct ra_ctx_fns ra_ctx_x11_egl;
+extern const struct ra_ctx_fns ra_ctx_drm_egl;
+extern const struct ra_ctx_fns ra_ctx_cocoa;
+extern const struct ra_ctx_fns ra_ctx_wayland_egl;
+extern const struct ra_ctx_fns ra_ctx_wgl;
+extern const struct ra_ctx_fns ra_ctx_angle;
+extern const struct ra_ctx_fns ra_ctx_dxinterop;
+extern const struct ra_ctx_fns ra_ctx_rpi;
+extern const struct ra_ctx_fns ra_ctx_mali;
+extern const struct ra_ctx_fns ra_ctx_vdpauglx;
+
+static const struct ra_ctx_fns *contexts[] = {
+// OpenGL contexts:
+#if HAVE_RPI
+    &ra_ctx_rpi,
+#endif
+/*
+#if HAVE_GL_COCOA
+    &ra_ctx_cocoa,
+#endif
+#if HAVE_EGL_ANGLE_WIN32
+    &ra_ctx_angle,
+#endif
+#if HAVE_GL_WIN32
+    &ra_ctx_wgl,
+#endif
+#if HAVE_GL_DXINTEROP
+    &ra_ctx_dxinterop,
+#endif
+*/
+#if HAVE_GL_X11
+    &ra_ctx_glx_probe,
+#endif
+#if HAVE_EGL_X11
+    &ra_ctx_x11_egl,
+#endif
+#if HAVE_GL_X11
+    &ra_ctx_glx,
+#endif
+#if HAVE_GL_WAYLAND
+    &ra_ctx_wayland_egl,
+#endif
+#if HAVE_EGL_DRM
+    &ra_ctx_drm_egl,
+#endif
+#if HAVE_MALI_FBDEV
+    &ra_ctx_mali,
+#endif
+#if HAVE_VDPAU_GL_X11
+    &ra_ctx_vdpauglx,
+#endif
+};
+
+static const char *type_names[] = {
+    [RA_API_OPENGL] = "opengl",
+};
```

Comment:
I'm not really fine with enumerating API names while we have generic backend things. That would mean adding a new API would require adding a new backend _and_ an enum item.

Reply:
We could just remove `--gpu-api` but I think users expect to be able to directly control this. Basically, I expect this option to directly map to the `ra_vk` implementation that we will get from it. So we will extend it every time we add a new ra_vk, which I really don't see being that frequent (personally).

Reply:
Can't you just use strings or so?

Reply:
I'm not sure what you mean? Instead of writing `.type = RA_API_OPENGL` you want us to type `.type = "opengl"`? What does that accomplish except making the code more prone to accidental typos? (And, of course, to avoid that I would personally rather `#define RA_API_OPENGL "opengl"` at that point)

Is extending an enum when you want to really that hard? Do you need to pay some patent fees or something? (Seriously, I don't get it)

Reply:
The common code just shouldn't know, have to know, or be able to know the API. Having a bunch of places that need to be updated when a backend is added is no justification for adding even more places (rather we should remove them). Ideally the manpage should be the only "other" place to update, and only 1. because the manpage can't be generated from the source, and 2. only if we really want to list all choices explicitly on the manpage (instead of letting "help" list it automatically).

Adding such redundancy would be a step back into MPlayer times, where some components used to have an extra enum to identify them.

It doesn't have to be a string (though in the end you need a string somewhere anyway). You could for example use a unique per-ra backend pointer (maybe `ra_fns`?). But in general I think this is the result of the misdesign that ra_context like MPGLContext conflates a number of mechanisms and thus causes a combinatorial explosion, instead of having orthogonal components which could be naturally selected separately.

Reply:
So do I understand it that you would be happier if we got rid of this enum, and handled `--gpu-api` like `--gpu-context`? I guess that's consistent, if nothing else.

> instead of having orthogonal components which could be naturally selected separately.

Is it really a combinatorial explosion when, as far as I can tell, every single context has unique code that's dependent on specifically that context? Even shit like “x11egl” that technically just combines “egl” (from egl_helpers) with “x11” (from x11.h) in the ismplest possible way has specific EGL/X11 interop code like `pick_xrgba_config`. If you tried separating “egl” from “x11”, where would this x11egl combination code live? All of the other contexts share a similar story: xlib/vulkan needs xlib-specific vulkan code, x11/opengl needs specific GLX code, dxinterop and angle need specific interop code, etc. etc. This code *HAS* to go somewhere, so we *HAVE* to have an exponential number of “combination X/Y/Z” files that contain this code.

It's a combinatorial explosion, perhaps, but it's one WE are powerless to do anything about. The explosion happens in the sheer number underlying APIs we're forced to use. Basicaly only vulkan allows this stuff to be orthogonally separated, but unless you want to drop support for opengl (which I'd vote in favor of, btw)...

If anything, I'd also like to try and get rid of `struct x11` etc. from `struct vo`, and make this live inside the `struct priv` for that specific combination.

Reply:
You're right about EGL etc., but not when it comes to things like GUI backends. For example how do you ask for "give me anything X11"? Will you add yet another enum? You could move the GUI backends (X11 etc.) to a separate list of "drivers", and then you could easily select them from a list. x11egl could be simply auto-selected from wanting OpenGL + X11. EGL vs. GLX is of course a fucked up details (just like all these Windows interop things and ANGLE), but these you would just select directly if you wanted.

Anyway now you did it as "express every possible sub-combination combination as individual ra_context", and you're running into the expected problems.

> If anything, I'd also like to try and get rid of struct x11 etc. from struct vo, and make this live inside the struct priv for that specific combination.

I've always planned this. It's conceptually easy to do, but would require messing with all the UI backends _and_ all MGPLContexts (which is why MPGLContext sucks ass).

Reply:
> For example how do you ask for "give me anything X11"? Will you add yet another enum? You could move the GUI backends (X11 etc.) to a separate list of "drivers"

If we want this for user-reasons we could also add a `const char *` to the ra_fns, so each combination (e.g. X11+EGL+OpenGL) would basically define “x11”, “egl” and “opengl” in its ra_fns, and the user could request either --gpu-api=opengl or --gpu-context=egl or --gpu-windowing=x11 and have a good chance of getting either one of these.

But I think apart from a few special cases like dxinterop, angle, and x11egl; the “context” is already very closely tied to the “windowing” system, to such an extent that we basically already name the context based on the windowing system. At that point I even question the value of whether or not `--gpu-windowing=x11 --gpu-context=egl` is really that much better than `--gpu-context=x11egl` for these special cases.

If I try and come up with cases in which selecting the windowing system directly is something the user would want to do, the only case I come up with is controlling wayland vs x11, and the only thing that throws a wrench into that equation is `x11egl`, because it's the only example of a situation in which `--gpu-context=x11` wouldn't do what the user expects.

(It's also worth at least pointing out the fact that there's a trade-off between “context matrix sparseness” and “context matrix flexibility” here. If we add even more options, then the public API will start having an exponential number of possible combinations, most of which will simply be unsupported. It already makes no sense to write `--gpu-api=vulkan --gpu-context=rpi`, for example. But that's not such a big deal since we have the `=help` list of all valid combinations. I just thought it was worth thinking about)